### PR TITLE
Don't allocate until we know it's worth it

### LINKED ISSUE #296: `#[serde(untagged)]` doesn't work with `RawValue`
For reasons outlined [here](https://github.com/serde-rs/json/issues/497), `RawValue` can't be
used with untagged enums at the moment. 
This means we can't use an `SingleOrBatch` untagged enum to handle incoming requests that can be one of:
- single JSONRPC request, i.e. `{"jsonrpc":"2.0","method":"bar","id":1}`
- a batch of requests, i.e. `[{"jsonrpc":"2.0","method":"baz","id":1}, {"jsonrpc":"2.0","method":"baz","id":2}]`
- a single notification, i.e. `{"jsonrpc":"2.0","method":"bezz"}`
- a batch of notifications, i.e. `[{"jsonrpc":"2.0","method":"bezz"}, {"jsonrpc":"2.0","method":"bozz"}, {"jsonrpc":"2.0","method":"bizz"}]`

(We do not support notifications, but if we ever were to do that to be fully compliant with the spec, this problem would make things worse)

Instead we have to try each case individually: first the single request case, then the batch case and then one time more to check for the error. For the worst case – unparseable input – we make three calls to [`serde_json::from_slice`] which is pretty annoying.

There is a PoC fix for this in https://github.com/serde-rs/serde/pull/1354

Suggestion: revisit this when/if https://github.com/serde-rs/serde/pull/1354 lands.

### PR DESCRIPTION
**Pull request by dvdplm**
For http server, check first byte before allocating space for the body.

Also re-work error handling  and prefer sending JSON-RPC-style errors (JSON doc, spec'd error codes) combined with http status codes.

Alternative to #419 

Closes #296

### CODE REVIEW

File: utils/src/hyper_helpers.rs [hunk 1]
```diff
@@ -29,24 +29,39 @@
...
 
-/// Read a hyper response with configured `HTTP` settings.
+/// Read a data from a [`hyper::Body`] and return the data if it is valid and within the allowed size range.
 ///
-/// Returns `Ok(bytes)` if the body was in valid size range.
+/// Returns `Ok((bytes, single))` if the body was in valid size range; and a bool indicating whether the JSON-RPC
+/// request is a single or a batch.
 /// Returns `Err` if the body was too large or the body couldn't be read.
-pub async fn read_response_to_body(
+pub async fn read_body(
```
Comment (dvdplm) [inline utils/src/hyper_helpers.rs:37]: The relevant changes are in this function.

File: http-server/src/server.rs [hunk 2]
```diff
@@ -201,61 +201,61 @@ impl Server {
 						}
 
 						let (parts, body) = request.into_parts();
-						let body = match read_response_to_body(&parts.headers, body, max_request_body_size).await {
-							Ok(body) => body,
-							Err(GenericTransportError::TooLarge) => {
-								return Ok::<_, HyperError>(response::too_large("The request was too large"))
-							}
+
+						let (body, mut is_single) = match read_body(&parts.headers, body, max_request_body_size).await {
```
Comment (dvdplm) [inline http-server/src/server.rs:205]: This is the relevant change of the PR.

File: utils/src/hyper_helpers.rs [hunk 3]
```diff
@@ -29,24 +29,39 @@
...
 	}
 
+	let first_chunk =
+		body.next().await.ok_or(GenericTransportError::Malformed)?.map_err(GenericTransportError::Inner)?;
+
+	if first_chunk.len() > max_request_body_size as usize {
+		return Err(GenericTransportError::TooLarge);
+	}
+
+	let single = match first_chunk[0] {
```
[the remark this answers is not in the document]
Comment (dvdplm) [reply utils/src/hyper_helpers.rs:57]: I _think_ `body.next().await` will return `None` for empty bodies (and become a `GenericTransportError::Malformed` error to us), but better to use `.get` anyway.

File: http-server/src/tests.rs [hunk 4]
```diff
@@ -199,7 +199,7 @@ async fn async_method_call_with_ok_context() {
 }
 
 #[tokio::test]
-async fn valid_batched_method_calls() {
+async fn valid_batched_method_callsx() {
```
Comment (niklasad1) [inline http-server/src/tests.rs:202]: hmm, is this change intended?

File: ws-server/src/tests.rs [hunk 5]
```diff
@@ -218,6 +218,7 @@ async fn can_set_max_connections() {
 
 #[tokio::test]
 async fn single_method_calls_works() {
+	println!("[test] START single");
```
Comment (niklasad1) [inline ws-server/src/tests.rs:221]: make these `log!` or remove?!

**Review by niklasad1 — APPROVED** [review]

**Pull request merged**